### PR TITLE
Fix for exact event values in the event value expression

### PR DESCRIPTION
**Pull request by TheLimeGlass**
### Description
With the new addition of extension classes to Entity in the form of EventValueExpression, we're seeing issues where Skript was preferring the entity over the egg in the provided related issue. "What came first? The egg or the entity?" Well Skript thought it was the Entity, but it was actually wrong ;)

So this pull request adds a boolean argument to the constructor of EventValueExpression to strictly tell the event value getter lookup to not check for subclasses and also to not convert.

A similar issue was present with ClassInfos back when Njol was making Skript, which is why we have so many methods in the Classes class https://github.com/SkriptLang/Skript/blob/1f20b98f5a7476b2a1caec298f72b3e360c170f7/src/main/java/ch/njol/skript/registrations/Classes.java#L291

Classes like Egg that extend Entity should not be grabbing Entity from the event. They should be only grabbing Egg and nothing else. If Egg doesn't exist, it should state that Egg doesn't exist in that event.

---
**Target Minecraft Versions:** any
**Requirements:** none
**Related Issues:** https://github.com/SkriptLang/Skript/issues/5539

### CODE REVIEW

File: src/main/java/ch/njol/skript/registrations/EventValues.java [hunk 1]
```diff
@@ -351,7 +372,11 @@ public T get(E e) {
 			}
 		};
 	}
-	
+
+	public static boolean doesExactEventValueHaveTimeStates(Class<? extends Event> e, Class<?> c) {
+		return getExactEventValueGetter(e, c, -1) != null || getExactEventValueGetter(e, c, 1) != null;
```
Comment (AyhamAl-Ali) [inline src/main/java/ch/njol/skript/registrations/EventValues.java:377]:
```suggestion
	public static boolean doesExactEventValueHaveTimeStates(Class<? extends Event> event, Class<?> c) {
		return getExactEventValueGetter(event, c, -1) != null || getExactEventValueGetter(event, c, 1) != null;
```

**Review by AyhamAl-Ali — APPROVED** [review]
Well done ⚡

File: src/main/java/ch/njol/skript/registrations/EventValues.java (hunk 1 above)
Comment (TheLimeGlass) [reply src/main/java/ch/njol/skript/registrations/EventValues.java:377]: Is done so in pull request https://github.com/SkriptLang/Skript/pull/4858

File: src/main/java/ch/njol/skript/registrations/EventValues.java [hunk 2]
```diff
@@ -351,9 +372,13 @@ public T get(E e) {
 			}
 		};
 	}
-	
-	public static boolean doesEventValueHaveTimeStates(Class<? extends Event> e, Class<?> c) {
-		return getEventValueGetter(e, c, -1, false) != null || getEventValueGetter(e, c, 1, false) != null;
+
+	public static boolean doesExactEventValueHaveTimeStates(Class<? extends Event> event, Class<?> c) {
+		return getExactEventValueGetter(event, c, -1) != null || getExactEventValueGetter(event, c, 1) != null;
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/registrations/EventValues.java:377]: Should use the time constants, right?

**Review by APickledWalrus — CHANGES_REQUESTED** [review]
Looks pretty good overall. Just a few minor things.

File: src/main/java/ch/njol/skript/expressions/base/PropertyExpression.java [hunk 3]
```diff
@@ -120,7 +120,7 @@ protected T[] get(final F[] source, final ch.njol.skript.classes.Converter<? sup
 	}
 
 	@Override
-	public final boolean isSingle() {
```
[the remark this answers is not in the document]
Comment (TheLimeGlass) [reply src/main/java/ch/njol/skript/expressions/base/PropertyExpression.java:123]: This was recently changed. Apon updating to master, this change will not be a change.

**Review by APickledWalrus — APPROVED** [review]

**Pull request merged**